hera_cal/red_groups.py

```diff
@@ -0,0 +1,407 @@
+        if isinstance(key, int):
+            self._red_list[key] = value
+        elif key in self:
+            ukey = self.get_ubl_key(key)
+            self._red_list = [value if red[0]==ukey else red for red in self._red_list]
+        else:
+            # We're setting a new redundant group
+            self.append(value)
+
+        # Reset the maps
+        self.clear_cache()
+
+    def index(self, key: AntPair | Baseline) -> int:
+        """Return the index of a redundant group."""
+        return list(self._red_key_to_bls_map.keys()).index(self.get_ubl_key(key))
+    
+    def filtered(self, **kwargs) -> RedundantGroups:
+        """Return a new RedundantGroups object with baselines filtered out.
+
+        """
+        new_reds = redcal.filter_reds(copy.deepcopy(self), **kwargs)
+        return attrs.evolve(self, red_list=new_reds)
+    
+    def extend(self, reds: Sequence[Sequence[AntPair | Baseline]], inplace: bool = True) -> None:
+        """In-place extend the list of redundant groups with another list of redundant groups.
+        
+        This maintains the list-of-lists duck-typing of the redundant groups.
+        """
+        for red in reds:
+            self.append(red, inplace)
+    
+    def sort(self):
+        """Sort the redundant groups in-place."""
+        self._red_list.sort()
+        self.clear_cache()
+
+    def clear_cache(self):
+        """Clear the cached maps."""
+        for att in (
+            self._bl_to_red_map,
+            self._red_key_to_bls_map,
+            self.data_ants,
+            self.data_bls,
+        ):
+            try:
+                del att
+            except AttributeError:
+                pass
+
+    def get_full_redundancies(
+        self, pols: Sequence[str] | None = None, 
+        include_autos: bool = True, 
+        bl_error_tol: float = 1.0,
+        pol_mode: str='1pol',
+    ) -> RedundantGroups:
+        """Create a RedundantGroups object from an antpos dictionary."""
+        if pols is None:
+            if len(next(iter(self._bl_to_red_map.values()))) == 2:
+                pols = ()
+            else:
+                pols = tuple({bl[2] for bl in self._red_key_to_bls_map.keys()})
+
+        if pols:
+            reds = redcal.get_reds(
+                self.antpos, 
+                pols=pols, # we end up removing the pols anyway.
+                bl_error_tol=bl_error_tol, 
+                include_autos=include_autos,
+                pol_mode=pol_mode,
+            )
+        else:
+            reds = redcal.get_pos_reds(
+                self.antpos, 
+                bl_error_tol=bl_error_tol, 
+                include_autos=include_autos
+            )
+
+        return attrs.evolve(self, red_list=reds)
+    
+    def keyed_on_bls(self, bls: Sequence[AntPair | Baseline]) -> RedundantGroups:
+        """Return a new RedundantGroups object keyed on the given baselines."""
```

Comment:
I'm not quite sure I follow the logic here... perhaps you can clarify in the docstring what happens if it can't find certain baselines or if a redundant group doesn't have a baseline in the list.

Reply:
OK I've given it a longer explanation. Basically the behaviour is the same as was happening in `RedDataContainer`.